### PR TITLE
Update ProGuard documentation

### LINKED ISSUE #1094: 1.0.0-RC2 Proguard Configuration Regression
**Describe the bug**
It appears that the move of the JSON format  from `core` to `json` with the update from `1.0.0-RC1` to `1.0.0-RC2` is causing the serializers for `JsonElement` to not be kept when minify is enabled with the current recommended Proguard configuration on the README.

**To Reproduce**
Attempt to encode a `JsonElement` to string:

```kotlin
Json {}.encodeToString(
    JsonElement.serializer(),
    buildJsonObject {
        put("key", JsonPrimitive("value"))
    }
)
```

When minify is enabled with this Proguard configuration (from the README):
```
-keepattributes *Annotation*, InnerClasses
-dontnote kotlinx.serialization.SerializationKt
-keep,includedescriptorclasses class com.alexvanyo.kotlintest.**$$serializer { *; } # <-- change package name to your app's
-keepclassmembers class com.alexvanyo.kotlintest.** { # <-- change package name to your app's
    *** Companion;
}
-keepclasseswithmembers class com.alexvanyo.kotlintest.** { # <-- change package name to your app's
    kotlinx.serialization.KSerializer serializer(...);
}
```
the app crashes:

```
2020-09-22 17:18:53.511 16848-16848/com.alexvanyo.kotlintest E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.alexvanyo.kotlintest, PID: 16848
    java.lang.ExceptionInInitializerError
        at d.a.n.r.<clinit>(JsonElementSerializers.kt:153)
        at d.a.n.g.c(JsonElementSerializers.kt:45)
        at d.a.n.g.b(JsonElementSerializers.kt:30)
        at d.a.n.u.h.g(StreamingJsonEncoder.kt:16)
        at d.a.n.a.a(Json.kt:73)
        at com.alexvanyo.kotlintest.MainActivity.onCreate(MainActivity.kt:107)
        at android.app.Activity.performCreate(Activity.java:8000)
        at android.app.Activity.performCreate(Activity.java:7984)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3422)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3601)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: d.a.f: Serializer for class 'e' is not found.
    Mark the class as @Serializable or provide the serializer explicitly.
        at d.a.m.n0.c(Platform.common.kt:104)
        at d.a.m.m0.k(Platform.kt:29)
        at d.a.i.d(Serializers.kt:56)
        at d.a.h.b(Unknown Source:1)
        at d.a.i.a(Serializers.kt:76)
        at d.a.i.e(Serializers.kt:66)
        at d.a.i.c(Serializers.kt:38)
        at d.a.h.a(Unknown Source:1)
        at d.a.n.r$a.<init>(JsonElementSerializers.kt:21)
        at d.a.n.r$a.<clinit>(JsonElementSerializers.kt:148)
        	... 21 more
```

**Expected behavior**
The app should not crash when minifying with the recommended Proguard configuration, as was the case with `1.0.0-RC1`.

On a hunch, applying the same Proguard format to `kotlinx.serialization.json` appears to fix the issue:
```
-keep,includedescriptorclasses class kotlinx.serialization.json.**$$serializer { *; } # <-- change package name to your app's
-keepclassmembers class kotlinx.serialization.json.** {
    *** Companion;
}
-keepclasseswithmembers class kotlinx.serialization.json.** { # <-- change package name to your app's
    kotlinx.serialization.KSerializer serializer(...);
}
```

If this is intended, I think the the recommended Proguard configuration on the README should be updated to include keeping the necessary classes from `kotlinx.serialization.json`.

**Environment**
 - Kotlin version: 1.4.10
 - Library version: 1.0.0-RC2
 - Kotlin platforms: Kotlin/JVM (Android)
 - Gradle version: 6.5.1
 - IDE version: Android Studio 4.0.1

### PR DESCRIPTION
**Pull request by qwwdfsad**
Fixes #1094

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -173,7 +173,8 @@ to add this to your `proguard-rules.pro`:
 
 ```proguard
 -keepattributes *Annotation*, InnerClasses
--dontnote kotlinx.serialization.SerializationKt
+-dontnote kotlinx.serialization.AnnotationsKt # core serialization annotations
+-keep,includedescriptorclasses class kotlinx.serialization.json.**$$serializer { *; } # JsonElement serializers
```
Comment (alexvanyo) [inline README.md:177]: With the goal of keeping `core` format agnostic, would it make sense to move this `json`-specific rule to its own section at the end of the `core` rules?

Comment (sandwwraith) [reply README.md:177]: Makes sense. @qwwdfsad You can also add commentary like 'you can omit this line if you aren't using Json format'

Comment (nabrozidhs) [inline README.md:177]: Unless I misunderstood something these two lines did not work for me as I kept having this issue:
```
Non-fatal Exception: java.lang.NoClassDefFoundError
kotlinx.serialization.json.JsonObjectSerializer
```

What I'm using and it seems to work in production now is this:
```
-keepattributes *Annotation*, InnerClasses
-dontnote kotlinx.serialization.AnnotationsKt # core serialization annotations
-keep,includedescriptorclasses class kotlinx.serialization.json.** { *; }
-keep,includedescriptorclasses class <APP PACKAGE NAME>.**$$serializer { *; }
-keepclassmembers class <APP PACKAGE NAME>.** {
    *** Companion;
}
-keepclasseswithmembers class <APP PACKAGE NAME>.** {
    kotlinx.serialization.KSerializer serializer(...);
}
```

The line `-keep,includedescriptorclasses class kotlinx.serialization.json.** { *; }` did the fix for me but I understand it's probably too broad.

Comment (igormalytsky) [reply README.md:177]: The same here.

Comment (qwwdfsad) [reply README.md:177]: Probably should be `-keep,includedescriptorclasses class kotlinx.serialization.json.**Serializer { *; }`

Comment (nabrozidhs) [reply README.md]: I'll give it a go now and let you know. My fix was more of a panic fix given it ended up in production :sweat_smile:

Comment (nabrozidhs) [reply README.md]: I'm still getting the java.lang.ExceptionInInitializerError
```
 Caused by: kotlinx.serialization.SerializationException: Serializer for class 'JsonElement' is not found.
Mark the class as @Serializable or provide the serializer explicitly.
    at kotlinx.serialization.internal.Platform_commonKt.serializerNotRegistered(Platform.common.kt:104)
    at kotlinx.serialization.internal.PlatformKt.platformSpecificSerializerNotRegistered(Platform.kt:29)
    at kotlinx.serialization.SerializersKt__SerializersKt.serializer(Serializers.kt:56)
    at kotlinx.serialization.SerializersKt.serializer(Unknown Source:1)
    at kotlinx.serialization.SerializersKt__SerializersKt.builtinSerializerOrNull$SerializersKt__SerializersKt(Serializers.kt:76)
    at kotlinx.serialization.SerializersKt__SerializersKt.serializerByKTypeImpl$SerializersKt__SerializersKt(Serializers.kt:66)
    at kotlinx.serialization.SerializersKt__SerializersKt.serializer(Serializers.kt:38)
    at kotlinx.serialization.SerializersKt.serializer(Unknown Source:1)
    at kotlinx.serialization.json.JsonObjectSerializer$JsonObjectDescriptor.<init>(JsonElementSerializers.kt:21)
    at kotlinx.serialization.json.JsonObjectSerializer$JsonObjectDescriptor.<clinit>(JsonElementSerializers.kt:148)
```

This is the setup after the change:
```
-keepattributes *Annotation*, InnerClasses
-dontnote kotlinx.serialization.AnnotationsKt # core serialization annotations
-keep,includedescriptorclasses class kotlinx.serialization.json.**$$serializer { *; }
-keep,includedescriptorclasses class kotlinx.serialization.json.**Serializer { *; }
-keep,includedescriptorclasses class <PACKAGE_NAME>.**$$serializer { *; }
-keepclassmembers class <PACKAGE_NAME>.** {
    *** Companion;
}
-keepclasseswithmembers class <PACKAGE_NAME>.** {
    kotlinx.serialization.KSerializer serializer(...);
}
```

**Pull request merged**